File: csharp/ToolGood.FlowVision/Common/RCY.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToolGood.FlowVision.Common
{
	/// <summary>
	/// RCY algorithm
	/// Author : Lin ZhiJun
	/// NickName : ToolGood
	/// Email : toolgood@qq.com
	/// 
	/// See https://github.com/toolgood/RCY
	/// </summary>
	static class RCY
	{
		private const int keyLen = 256;

		/// <summary>
		/// Encrypt
		/// </summary>
		/// <param name="data"></param>
		/// <param name="pass"></param>
		/// <returns></returns>
		public static byte[] Encrypt(byte[] data, byte[] pass)
		{
			if (data == null) throw new ArgumentNullException("data");
			if (data.Length == 0) throw new ArgumentNullException("data");
			if (pass == null) throw new ArgumentNullException("pass");
			if (pass.Length == 0) throw new ArgumentNullException("pass");

			return encrypt(data, pass);
		}
		private unsafe static byte[] encrypt(byte[] data, byte[] pass)
		{
			byte[] mBox = GetKey(pass, keyLen);
			byte[] output = new byte[data.Length];
			int i = 0, j = 0;
			var length = data.Length;


			fixed (byte* _mBox = &mBox[0])
			fixed (byte* _data = &data[0])
			fixed (byte* _output = &output[0]) {
				for (long offset = 0; offset < length; offset++) {
					i = ++i & 0xFF;
					j = j + *(_mBox + i) & 0xFF;

					byte a = *(_data + offset);
					byte c = (byte)(a ^ *(_mBox + (*(_mBox + i) & *(_mBox + j))));
					*(_output + offset) = c;

					j = j + a + c;
				}
			}

			return output;
		}



		private static unsafe byte[] GetKey(byte[] pass, int kLen)
		{
			byte[] mBox = new byte[kLen];
			fixed (byte* _mBox = &mBox[0]) {
				for (long i = 0; i < kLen; i++) {
					*(_mBox + i) = (byte)i;
				}
				long j = 0;
				int lengh = pass.Length;
				fixed (byte* _pass = &pass[0]) {
					for (long i = 0; i < kLen; i++) {
						j = (j + *(_mBox + i) + *(_pass + i % lengh)) % kLen;
						byte temp = *(_mBox + i);
						*(_mBox + i) = *(_mBox + j);
						*(_mBox + j) = temp;
					}
				}
			}
			return mBox;
		}
	}
}

File: java/src/main/java/toolgood/flowVision/Common/RCY.java
package toolgood.flowVision.Common;

public class RCY {
    private static final int keyLen = 256;

    public static byte[] Encrypt(byte[] data, byte[] pass) throws Exception {
        if (data == null) throw new Exception("data");
        if (data.length == 0) throw new Exception("data");
        if (pass == null) throw new Exception("pass");
        if (pass.length == 0) throw new Exception("pass");

        return encrypt(data, pass);
    }

    private static byte[] encrypt(byte[] data, byte[] pass) {
        byte[] mBox = GetKey(pass, keyLen);
        byte[] output = new byte[data.length];
        int i = 0, j = 0;
        byte a, c;
        for (int offset = 0; offset < data.length; offset++) {
            i = (++i) & 0xFF;
            j = (j + (mBox[i]& 0xff)) & 0xFF;

            a = data[offset];
            c = (byte) (a ^ mBox[((mBox[i] & 0xff) & (mBox[j] & 0xff))]);
            output[offset] = c;

            j = j + (a& 0xff) + (c& 0xff);
        }
        return output;
    }


    private static byte[] GetKey(byte[] pass, int kLen) {
        byte[] mBox = new byte[kLen];

        for (int i = 0; i < kLen; i++) {
            mBox[i] = (byte) i;
        }
        int j = 0;
        for (int i = 0; i < kLen; i++) {
            j = (j + (mBox[i] & 0xff) + pass[i % pass.length]) % kLen;
            byte temp = mBox[i];
            mBox[i] = mBox[j];
            mBox[j] = temp;
        }
        return mBox;
    }
}
